File: Reqnroll/Chapter15/15.4-TheTestActionPattern/After/WIMP.Specs/Support/RestApiContext.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;

using WIMP.App.RestApi;

namespace WIMP.Specs.Support;

public record VoidReturn
{
    public static readonly VoidReturn Instance = new();
}

public class RestApiContext(AppHostingContext appHostingContext)
{
    public string? BearerToken { get; set; }

    public async Task<TResult> GetRequest<TResult>(
        string path, object? payload = null)
    {
        var response = await SendRequest(HttpMethod.Get, path, payload);
        response.EnsureSuccessStatusCode();
        return await ReadResult<TResult>(response);
    }

    public async Task<TResult> ProcessRequest<TResult>(
        string actionName, HttpMethod method, string path,
        object? payload = null,
        HttpStatusCode successStatusCode = HttpStatusCode.OK)
    {
        var response = await SendRequest(method, path, payload);
        if (response.StatusCode != successStatusCode)
        {
            string errorMessage = await ReadErrorMessage(response);
            throw new WimpActionFailedException(
                $"{actionName} failed with status code {response.StatusCode}. " +
                $"Error message: '{errorMessage}'");
        }

        return typeof(TResult) == typeof(VoidReturn)
            ? default!
            : await ReadResult<TResult>(response);
    }

    private async Task<HttpResponseMessage> SendRequest(
        HttpMethod method, string path, object? payload = null)
    {
        var request = new HttpRequestMessage(method, path);
        if (!string.IsNullOrEmpty(BearerToken))
        {
            request.Headers.Authorization =
                new AuthenticationHeaderValue("Bearer", BearerToken);
        }

        if (payload != null)
        {
            request.Content = JsonContent.Create(payload);
        }

        var httpClient = appHostingContext.AppHost.CreateClient();
        return await httpClient.SendAsync(request);
    }

    private async Task<TResult> ReadResult<TResult>(HttpResponseMessage response)
    {
        return await response.Content.ReadFromJsonAsync<TResult>()
            ?? throw new InvalidOperationException("No result payload found");
    }

    private async Task<string> ReadErrorMessage(HttpResponseMessage response)
    {
        string? errorMessage =
            (await response.Content.ReadFromJsonAsync<ErrorResponse>())?.Error;
        return string.IsNullOrEmpty(errorMessage) ? "n/a" : errorMessage;
    }
}

File: CucumberJVM/Chapter15/15.7-StubDependencyPattern/Before/src/test/java/com/wimp/specs/support/RestApiContext.java
package com.wimp.specs.support;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonSyntaxException;
import com.google.gson.JsonSerializer;
import com.wimp.app.restapi.ErrorResponse;
import io.cucumber.spring.ScenarioScope;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockHttpServletResponse;
import org.springframework.stereotype.Component;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

@Component
@ScenarioScope
public class RestApiContext {
    private final MockMvc mockMvc;
    private final Gson gson;
    private String bearerToken;

    public RestApiContext(MockMvc mockMvc) {
        this.mockMvc = mockMvc;
        this.gson = new GsonBuilder()
            .registerTypeAdapter(LocalDateTime.class, (JsonSerializer<LocalDateTime>) (src, typeOfSrc, context) ->
                context.serialize(src.format(DateTimeFormatter.ISO_LOCAL_DATE_TIME)))
            .registerTypeAdapter(LocalDateTime.class, (JsonDeserializer<LocalDateTime>) (json, typeOfT, context) ->
                LocalDateTime.parse(json.getAsString(), DateTimeFormatter.ISO_LOCAL_DATE_TIME))
            .create();
    }

    public void setBearerToken(String bearerToken) {
        this.bearerToken = bearerToken;
    }

    public void clearBearerToken() {
        this.bearerToken = null;
    }

    public <TResult> TResult getRequest(String path, Class<TResult> responseType) throws Exception {
        return processRequest("GET " + path, HttpMethod.GET, path, null, HttpStatus.OK, responseType);
    }

    public <TResult> TResult processRequest(
        String actionName,
        HttpMethod method,
        String path,
        Object payload,
        HttpStatus successStatusCode,
        Class<TResult> responseType
    ) throws Exception {
        MockHttpServletRequestBuilder requestBuilder = MockMvcRequestBuilders.request(method, path);
        if (bearerToken != null && !bearerToken.isBlank()) {
            requestBuilder.header(HttpHeaders.AUTHORIZATION, "Bearer " + bearerToken);
        }
        if (payload != null) {
            requestBuilder.contentType(MediaType.APPLICATION_JSON)
                .content(gson.toJson(payload));
        }

        MockHttpServletResponse response = mockMvc.perform(requestBuilder).andReturn().getResponse();
        if (response.getStatus() != successStatusCode.value()) {
            throw new TestActionFailedException(actionName + " failed with status code " + response.getStatus()
                + ". Error message: '" + readErrorMessage(response) + "'");
        }

        if (responseType == VoidReturn.class) {
            return responseType.cast(VoidReturn.INSTANCE);
        }

        return gson.fromJson(response.getContentAsString(), responseType);
    }

    private String readErrorMessage(MockHttpServletResponse response) throws Exception {
        String responseText = response.getContentAsString();
        if (responseText == null || responseText.isBlank()) {
            return "n/a";
        }

        try {
            ErrorResponse errorResponse = gson.fromJson(responseText, ErrorResponse.class);
            if (errorResponse.error() != null && !errorResponse.error().isBlank()) {
                return errorResponse.error();
            }
        } catch (JsonSyntaxException ignored) {
            // ignored on purpose: if it is not a valid error payload, return raw response text
        }

        return responseText;
    }
}
